src/mjolnir/graphbuilder.cc

```diff
@@ -76,27 +78,74 @@ std::map<GraphId, size_t> SortGraph(const std::string& nodes_file, const std::st
       node.graph_id.set_id(last_node.graph_id.id());
     }
 
-    // if this node marks the start of an edge, go tell the edge where the first node in the
-    // series is
+    // if this node marks the start of an edge, keep track of the edge and the node
+    // so we can later tell the edge where the first node in the series is
     if (node.is_start()) {
-      auto element = edges[node.start_of];
-      auto edge = *element;
-      edge.sourcenode_ = run_index;
-      element = edge;
+      starts.emplace_back(node.start_of, run_index);
     }
     // if this node marks the end of an edge, go tell the edge where the first node in the
     // series is
     if (node.is_end()) {
-      auto element = edges[node.end_of];
-      auto edge = *element;
-      edge.targetnode_ = run_index;
-      element = edge;
+      ends.emplace_back(node.end_of, run_index);
     }
 
     // next node
     last_node = node;
     ++node_index;
   });
+  
+  LOG_INFO("Nodes processed. Now adding data to edges.");
```

Comment:
the rest of this looks perfectly fine. its unfortunate that looping over both sets at the same time ends up being so verbose but there isnt a clearer way to do it so it looks good to me!